pyintegration/deephaven2/table_listener.py

```diff
@@ -0,0 +1,270 @@
+
+        Raises:
+            RuntimeError
+        """
+
+        if self.isRegistered:
+            raise RuntimeError("Attempting to register an already registered listener..")
+
+        self.t.j_table.listenForUpdates(self.listener)
+
+        self.isRegistered = True
+
+    def deregister(self):
+        """ Deregister the listener from the table and stop listening for updates.
+
+        Raises:
+            RuntimeError
+        """
+
+        if not self.isRegistered:
+            raise RuntimeError("Attempting to deregister an unregistered listener..")
+
+        self.t.j_table.removeUpdateListener(self.listener)
+        self.isRegistered = False
+
+
+def _do_locked(f: Callable, lock_type="shared"):
+    """ Executes a function while holding the UpdateGraphProcessor (UGP) lock.  Holding the UGP lock
+    ensures that the contents of a table will not change during a computation, but holding
+    the lock also prevents table updates from happening.  The lock should be held for as little
+    time as possible.
+
+    Args:
+        f (Callable): callable to execute while holding the UGP lock, could be function or an object with an 'apply'
+            attribute which is callable
+        lock_type (str): UGP lock type, valid values are "exclusive" and "shared".  "exclusive" allows only a single
+            reader or writer to hold the lock.  "shared" allows multiple readers or a single writer to hold the lock.
+    Raises:
+        ValueError
+    """
+    throwing_runnable = jpy.get_type("io.deephaven.integrations.python.PythonThrowingRunnable")
+    update_graph_processor = jpy.get_type("io.deephaven.engine.updategraph.UpdateGraphProcessor")
+
+    if lock_type == "exclusive":
+        update_graph_processor.DEFAULT.exclusiveLock().doLocked(throwing_runnable(f))
+    elif lock_type == "shared":
+        update_graph_processor.DEFAULT.sharedLock().doLocked(throwing_runnable(f))
+    else:
+        raise ValueError(f"Unsupported lock type: lock_type={lock_type}")
+
+
+def _nargs_listener(listener) -> int:
+    """ Returns the number of arguments the listener takes.
+
+    Args:
+        listener: listener
+
+    Returns:
+        the number of arguments the listener takes
+
+    Raises:
+        ValueError, NotImplementedError
+    """
+
+    if callable(listener):
+        f = listener
+    elif hasattr(listener, 'onUpdate'):
+        f = listener.onUpdate
+    else:
+        raise ValueError("ShiftObliviousListener is neither callable nor has an 'onUpdate' method")
+
+    if sys.version[0] == "2":
+        import inspect
+        return len(inspect.getargspec(f).args)
+    elif sys.version[0] == "3":
+        from inspect import signature
+        return len(signature(f).parameters)
+    else:
+        raise NotImplementedError("Unsupported python version: version={}".format(sys.version))
+
+
+def listen(t: Table, listener: Callable, description: str = None, retain: bool = True, listener_type: str = "auto",
```

Comment:
Listener is either a callable or an object with an onUpdate method, so would `Union[Callable, object] be more appropriate?

Change also needed in docs.